File: PaperLib/Models/PaperEntity.swift
//
//  PaperEntity.swift
//  PaperLib
//
//  Created by GeoffreyChen on 26/11/2021.
//

import Foundation
import RealmSwift

class PaperEntityIDObject: Object {
    @objc dynamic var id: ObjectId
    @objc dynamic var _id: ObjectId

    override init() {
        self.id = ObjectId.generate()
        self._id = self.id
    }

    override class func primaryKey() -> String? {
        return "_id"
    }

    override class func shouldIncludeInDefaultSchema() -> Bool {
        self != PaperEntityIDObject.self
    }
}

class PaperEntity: PaperEntityIDObject, ObjectKeyIdentifiable {

    @Persisted var _partition: String?
    @Persisted var addTime: Date

    @Persisted var title: String = ""
    @Persisted var authors: String = ""
    @Persisted var publication: String = ""
    @Persisted var pubTime: String = ""
    @Persisted var pubType: Int = 2 // 0: Article; 1: Conference; 2: Others
    @Persisted var doi: String = ""
    @Persisted var arxiv: String = ""
    @Persisted var mainURL: String = ""
    @Persisted var supURLs: List<String>
    @Persisted var rating: Int = 0
    @Persisted var tags: List<PaperTag>
    @Persisted var folders: List<PaperFolder>
    @Persisted var flag: Bool = false
    @Persisted var note: String = ""

    convenience init(
        title: String?,
        authors: String?,
        publication: String?,
        pubTime: String?,
        pubType: Int?,
        doi: String?,
        arxiv: String?,
        mainURL: String?,
        supURLs: [String]?,
        rating: Int?,
        flag: Bool?
    ) {
        self.init()

        self.id = self._id

        self.addTime = Date()

        self.title = title ?? ""
        self.authors = authors ?? ""
        self.publication = publication ?? ""
        self.pubTime = pubTime ?? ""
        self.pubType = pubType ?? 2
        self.doi = doi ?? ""
        self.arxiv = arxiv ?? ""
        self.mainURL = mainURL ?? ""
        supURLs?.forEach { supURL in
            self.supURLs.append(supURL)
        }
        self.rating = rating ?? 0
        self.flag = flag ?? false

        print("create entity")
    }
}

class PaperEntityDraft: NSObject {
    @objc var id: ObjectId = .generate()
    @objc var addTime: Date = .init()

    @objc var title: String = ""
    @objc var authors: String = ""
    @objc var publication: String = ""
    @objc var pubTime: String = ""
    @objc var pubType: String = "Others" // 0: Article; 1: Conference; 2: Others
    @objc var doi: String = ""
    @objc var arxiv: String = ""
    @objc var mainURL: String = ""
    @objc var supURLs: [String] = .init()
    @objc var rating: Int = 0
    @objc var tags: String = ""
    @objc var folders: String = ""
    @objc var flag: Bool = false
    @objc var note: String = ""

    override init() {
        super.init()
    }

    init(from entity: PaperEntity) {
        id = entity.id
        addTime = entity.addTime

        title = entity.title
        authors = entity.authors
        publication = entity.publication
        pubTime = entity.pubTime
        pubType = ["Journal", "Conference", "Others"][entity.pubType]
        doi = entity.doi
        arxiv = entity.arxiv
        mainURL = entity.mainURL
        supURLs = Array(entity.supURLs)
        rating = entity.rating
        tags = Array(entity.tags.map { formatString($0.name, returnEmpty: true, removeStr: "tag-")! }).joined(separator: "; ")
        folders = Array(entity.folders.map { formatString($0.name, returnEmpty: true, removeStr: "folder-")! }).joined(separator: "; ")
        flag = entity.flag
        note = entity.note

    }

    func set(for key: String, value: Any?, allowEmpty: Bool = false) {
        guard value != nil || allowEmpty else { return }

        var formatedValue = value
        if formatedValue is String {
            if key == "title" || key == "authors" {
                formatedValue = formatString(formatedValue as? String, removeNewline: true, removeStr: ".")
            }
            if (formatedValue as? String ?? "").isEmpty {
                if allowEmpty {
                    setValue(formatedValue, forKey: key)
                }
                return
            } else {
                setValue(formatedValue, forKey: key)
            }
        } else {
            setValue(formatedValue, forKey: key)
            return
        }
    }
}

File: app/models/paper-entity.ts
import { ObjectId } from "bson";
import Mathml2latex from "mathml-to-latex";

import { ICategorizerDraft, PaperFolder, PaperTag } from "./categorizer";
import { FeedEntity } from "./feed-entity";
import { OID } from "./id";

export interface IPaperEntityDraft {
  _id?: OID;
  id?: OID;
  _partition?: string;
  addTime?: Date;
  title?: string;
  authors?: string;
  publication?: string;
  pubTime?: string;
  pubType?: number;
  doi?: string;
  arxiv?: string;
  mainURL?: string;
  supURLs?: string[];
  rating?: number;
  tags?: ICategorizerDraft[];
  folders?: ICategorizerDraft[];
  flag?: boolean;
  note?: string;
  codes?: string[];
  pages?: string;
  volume?: string;
  number?: string;
  publisher?: string;
}

export class PaperEntity {
  static schema = {
    name: "PaperEntity",
    primaryKey: "_id",
    properties: {
      id: "objectId",
      _id: "objectId",
      _partition: "string?",
      addTime: "date",

      title: "string",
      authors: "string",
      publication: "string",
      pubTime: "string",
      pubType: "int",
      doi: "string",
      arxiv: "string",
      mainURL: "string",
      supURLs: {
        type: "list",
        objectType: "string",
      },
      rating: "int",
      tags: {
        type: "list",
        objectType: "PaperTag",
      },
      folders: {
        type: "list",
        objectType: "PaperFolder",
      },
      flag: "bool",
      note: "string",
      codes: {
        type: "list",
        objectType: "string",
      },
      pages: "string",
      volume: "string",
      number: "string",
      publisher: "string",
    },
  };

  _id: OID;
  id: OID;
  _partition: string;
  addTime: Date;
  title: string;
  authors: string;
  publication: string;
  pubTime: string;
  pubType: number;
  doi: string;
  arxiv: string;
  mainURL: string;
  supURLs: string[];
  rating: number;
  tags: PaperTag[];
  folders: PaperFolder[];
  flag: boolean;
  note: string;
  codes: string[];
  pages: string;
  volume: string;
  number: string;
  publisher: string;

  constructor(object?: IPaperEntityDraft, initObjectId = false) {
    this._id = object?._id ? new ObjectId(object._id) : "";
    this.id = object?._id ? new ObjectId(object._id) : "";
    this.id = this.id ? this.id : this._id;
    this._id = this._id ? this._id : this.id;

    this._partition = object?._partition || "";
    this.addTime = object?.addTime || new Date();
    this.title = object?.title || "";
    this.authors = object?.authors || "";
    this.publication = object?.publication || "";
    this.pubTime = object?.pubTime || "";
    this.pubType = object?.pubType || 0;
    this.doi = object?.doi || "";
    this.arxiv = object?.arxiv || "";
    this.mainURL = object?.mainURL || "";
    this.supURLs = object?.supURLs?.map((url) => url) || [];
    this.rating = object?.rating || 0;
    this.tags = object?.tags?.map((tag) => new PaperTag(tag, false)) || [];
    this.folders =
      object?.folders?.map((folder) => new PaperFolder(folder, false)) || [];
    this.flag = object?.flag || false;
    this.note = object?.note || "";
    this.codes = object?.codes?.map((code) => code) || [];
    this.pages = object?.pages || "";
    this.volume = object?.volume || "";
    this.number = object?.number || "";
    this.publisher = object?.publisher || "";

    if (initObjectId) {
      this._id = new ObjectId();
      this.id = this._id;
    }

    return new Proxy(this, {
      set: (target, prop, value) => {
        if (prop === "title") {
          target.setValue("title", value, true);
        } else if ((prop === "_id" || prop === "id") && value) {
          this._id = new ObjectId(value);
          this.id = this._id;
        } else {
          target[prop] = value;
        }

        return true;
      },
    });
  }

  setValue(key: keyof PaperEntity, value: unknown, format = false) {
    // Format the value
    if (format && value) {
      // Check if contains Mathml
      const mathmlRegex1 = /<math\b[^>]*>([\s\S]*?)<\/math>/gm;
      const mathmlRegex2 = /<mml:math\b[^>]*>([\s\S]*?)<\/mml:math>/gm;
      const mathmlRegex3 = /<mrow\b[^>]*>([\s\S]*?)<\/mrow>/gm;

      for (const regex of [mathmlRegex1, mathmlRegex2, mathmlRegex3]) {
        if (regex.test(value as string)) {
          const mathmls = (value as string).match(regex);
          if (mathmls) {
            for (const mathml of mathmls) {
              const latex = Mathml2latex.convert(mathml.replaceAll("mml:", ""));
              value = (value as string).replace(mathml, "$" + latex + "$");
            }
          }
        }
      }
    }
    this[key as any] = value;
  }

  initialize(object: IPaperEntityDraft) {
    this._id = object._id ? new ObjectId(object._id) : "";
    this.id = object._id ? new ObjectId(object._id) : "";
    this.id = this.id ? this.id : this._id;
    this._id = this._id ? this._id : this.id;

    this._partition = object._partition || "";
    this.addTime = object.addTime || new Date();
    this.title = object.title || "";
    this.authors = object.authors || "";
    this.publication = object.publication || "";
    this.pubTime = object.pubTime || "";
    this.pubType = object.pubType || 0;
    this.doi = object.doi || "";
    this.arxiv = object.arxiv || "";
    this.mainURL = object.mainURL || "";
    this.supURLs = object.supURLs || [];
    this.rating = object.rating || 0;
    this.tags = object.tags?.map((tag) => new PaperTag(tag, false)) || [];
    this.folders =
      object.folders?.map((folder) => new PaperFolder(folder, false)) || [];
    this.flag = object.flag || false;
    this.note = object.note || "";
    this.codes = object.codes || [];
    this.pages = object.pages || "";
    this.volume = object.volume || "";
    this.number = object.number || "";
    this.publisher = object.publisher || "";

    return this;
  }

  fromFeed(feedEntity: FeedEntity) {
    this.title = feedEntity.title;
    this.authors = feedEntity.authors;
    this.publication = feedEntity.publication;
    this.pubTime = feedEntity.pubTime;
    this.pubType = feedEntity.pubType;
    this.doi = feedEntity.doi;
    this.arxiv = feedEntity.arxiv;
    this.mainURL = feedEntity.mainURL;
    this.pages = feedEntity.pages;
    this.volume = feedEntity.volume;
    this.number = feedEntity.number;
    this.publisher = feedEntity.publisher;
    return this;
  }
}
